File: programming_languages/Java/code/T51/test_T51.java
import java.util.Arrays;
import java.util.Random;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.IOException;

public class test_T51 {

    public static void main(String[] args) {
        test_files_identical();
        test_identical_output_check_all();
    }

    public static void test_files_identical() {
        String[][] filesToCompare = {
            {"Test1/test.java", "Test1_Xint/test.java"},
            {"Test2/test.java", "Test2_Xint/test.java"}
        };

        for (String[] pair : filesToCompare) {
            try {
                Path pathOrig = Paths.get(pair[0]);
                Path pathXint = Paths.get(pair[1]);

                if (!Files.exists(pathOrig) || !Files.exists(pathXint)) {
                    System.err.println("[ERROR] File not found: " + pair[0] + " or " + pair[1]);
                    System.exit(1);
                }

                java.util.List<String> f1Lines = Files.readAllLines(pathOrig);
                java.util.List<String> f2Lines = Files.readAllLines(pathXint);

                f1Lines.removeIf(line -> ignoreLine(line));
                f2Lines.removeIf(line -> ignoreLine(line));

                if (!f1Lines.equals(f2Lines)) {
                    System.err.println("[ERROR] Files differ (ignoring package & REPS): "
                            + pair[0] + " != " + pair[1]);
                    System.exit(1);
                }

            } catch (IOException e) {
                System.err.println("[ERROR] Error reading files: " + e.getMessage());
                System.exit(1);
            }
        }
    }

    private static boolean ignoreLine(String line) {
        String l = line.trim();

        return l.startsWith("package ")
            || l.startsWith("public static final int REPS")
            || l.startsWith("public static final long REPS")
            || l.contains("for (int reps")
            || l.contains("for (long reps");
    }

    public static void test_identical_output_check_all() {
        try {
            Random random = new Random();

            int n = Test1.test.N;

            int[] masterArray = new int[n];
            for (int i = 0; i < n; i++) {
                masterArray[i] = random.nextInt(n);
            }

            int[] copy1 = Arrays.copyOf(masterArray, n);
            int[] copy2 = Arrays.copyOf(masterArray, n);

            Test1.test.test(copy1);
            Test2.test.test(copy2);

            if (!Arrays.equals(copy1, copy2)) {
                System.err.println("[ERROR] Mismatch detected between versions!");
                System.exit(1);
            }

            int[] sortedMaster = Arrays.copyOf(masterArray, n);
            Arrays.sort(sortedMaster);

            if (!Arrays.equals(copy1, sortedMaster)) {
                System.err.println("[ERROR] The arrays are identical but they are NOT sorted!");
                System.exit(1);
            }

        } catch (Throwable e) {
            System.err.println("[ERROR] Execution error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: programming_languages/Python/code/T51/test_T51.py
import os
import random
import Test1.test as file1
import Test2.test as file2
from Test1.test import N

files_to_compare = [
    ("Test1/test.py", "Test1_Nuitka/test.py"),
    ("Test2/test.py", "Test2_Nuitka/test.py")
]

def test_files_identical():
    for orig, nuitka in files_to_compare:
        assert os.path.exists(orig), f"Original file not found: {orig}"
        assert os.path.exists(nuitka), f"Nuitka file not found: {nuitka}"

        with open(orig, "r", encoding="utf-8") as f1:
            content1 = f1.read()
        with open(nuitka, "r", encoding="utf-8") as f2:
            content2 = f2.read()

        assert content1 == content2, f"Files differ: {orig} != {nuitka}"

def test_identical_output_check_all():
    master_array = [random.randint(0, N - 1) for _ in range(N)]

    copy1 = list(master_array)
    copy2 = list(master_array)

    file1.test(copy1)
    file2.test(copy2)

    results = [copy1, copy2]
    names = ["File1", "File1_Nuitka", "File2", "File2_Nuitka"]

    for i, res in enumerate(results):
        if not (res == sorted(res)):
            raise AssertionError(f"Error: {names[i]} output is NOT sorted!")
        
    if not (copy1 == copy2):
            raise AssertionError(f"Mismatch detected between versions!")

if __name__ == "__main__":
    test_files_identical()
    test_identical_output_check_all()